### PR TITLE
Fix mem leaks and avoid NULL ptr deref

### PR DESCRIPTION
**Pull request by sander-visser**
And add NULL ptr deref pretection etc.

### CODE REVIEW

File: core/arch/arm/kernel/ree_fs_ta.c [hunk 1]
```diff
@@ -229,8 +229,10 @@ static TEE_Result ree_fs_ta_open(const TEE_UUID *uuid,
 		memcpy(&img_ehdr, ((uint8_t *)ta + offs), sizeof(img_ehdr));
 
 		ehdr = malloc(SHDR_ENC_GET_SIZE(&img_ehdr));
-		if (!ehdr)
-			return TEE_ERROR_OUT_OF_MEMORY;
+		if (!ehdr) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto error_free_handle;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/ree_fs_ta.c:234]: good catch. but this should rather be `goto error_free_hash;` (and label `error_free_handle` is not needed).

File: core/arch/arm/mm/tee_mmu.c [hunk 2]
```diff
@@ -370,8 +370,10 @@ static TEE_Result split_vm_region(struct user_mode_ctx *uctx,
 	if (mobj_is_paged(r->mobj)) {
 		TEE_Result res = tee_pager_split_um_region(uctx, va);
 
-		if (res)
+		if (res != TEE_SUCCESS) {
+			free(r2);
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_mmu.c:374]: Good catch also.
Here, prefer keep `if (res)` instead of `if (res != TEE_SUCCESS)`.

File: core/arch/arm/mm/tee_pager.c [hunk 3]
```diff
@@ -532,9 +532,14 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
...
 	size_t fobj_pgoffs = 0;
 	vaddr_t b = base;
-	size_t s = fobj->num_pages * SMALL_PAGE_SIZE;
+	size_t s;
 	size_t s2 = 0;
 
+	if (!fobj)
+		panic();
+
+	s = fobj->num_pages * SMALL_PAGE_SIZE;
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:541]: This change is not really needed, if `fobj` is NULL, Core would already panic on a data abort (seg fault).
Moreover, we don't expected a NULL `fobj` at entry.
Maybe we can use `assert(fobj);`, not `if (fobj) panic();` here, not nicely catch the issue in debug mode but I don't think it would add much.
I suggest to discard this change.

File: core/crypto/aes-gcm-sw.c [hunk 4]
```diff
@@ -43,6 +43,8 @@ void internal_aes_gcm_ghash_update(struct internal_aes_gcm_state *state,
 	if (head)
 		ghash_update_block(state, head);
 
+	assert((num_blocks == 0) || (data != NULL));
```
Comment (etienne-lms) [inline core/crypto/aes-gcm-sw.c:46]: I'm not sure this is really needed, but not a strong opinion.
This may be useful only for those who hack into AES GCM software implementation.

Note: with optee_os coding style, this should rather be `assert(!num_blocks || data);`.

File: core/drivers/crypto/crypto_api/acipher/rsa.c [hunk 5]
```diff
@@ -102,7 +102,8 @@ TEE_Result crypto_acipher_rsanopad_decrypt(struct rsa_keypair *key,
 		CRYPTO_TRACE("Parameters error (key @%p)\n"
 			     "(msg @%p size %zu bytes)\n"
 			     "(cipher @0%p size %zu bytes)",
-			     key, msg, *msg_len, cipher, cipher_len);
+			     key, msg, msg_len != NULL ? *msg_len : 0,
```
Comment (etienne-lms) [inline core/drivers/crypto/crypto_api/acipher/rsa.c:105]: Good catch!
Prefer `msg_len ? *msg_len : 0` over `msg_len != NULL ? *msg_len : 0`.
Also, I think , when `msg_len` is NULL, size could be printed as a specific unexpected value, maybe `~0`.

This comment also applies to the other changes in this patch.

File: core/crypto/aes-gcm-sw.c (hunk 4 above)
Comment (jenswi-linaro) [reply core/crypto/aes-gcm-sw.c:46]: This looks like a random place to put this assert. Why is this needed?

Comment (sander-visser) [reply core/crypto/aes-gcm-sw.c:46]: assert to make sure internal_aes_gcm_xor_block would not end up dereferencing data that could have been NULL

File: core/arch/arm/mm/tee_mmu.c (hunk 2 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_mmu.c:374]: So optee coding style allowing enums to be treated as essential bools?

Comment (etienne-lms) [reply core/arch/arm/mm/tee_mmu.c:374]: Yes, `if (foo)` can be used with `foo` being a boolean, a pointer or a numerical value (even  an enumerated value).

By definition `TEE_SUCCESS` is zero. That's in the GPD TEE spec OP-TEE is based on. Using `if (res)` instead of `if (res != TEE_SUCCESS)` and `if (!res)` instead of `if (res == TEE_SUCCESS)` make to code easier to read since we really often test `TEE_Result` value against `TEE_SUCCESS`.

Note: optee_os coding style is mostly based on the Linux kernel coding style.

File: core/crypto/aes-gcm-sw.c (hunk 4 above)
Comment (jenswi-linaro) [reply core/crypto/aes-gcm-sw.c:46]: I get that part. Why is it needed? What prompted this?

Comment (etienne-lms) [reply core/crypto/aes-gcm-sw.c:46]: Be caution: `assert()` has effects only in debug mode (`CFG_TEE_CORE_DEBUG=y` in optee_os).
Faulting assert will make Core to panic with a panic trace.
Dereferencing a NULL pointer will make Core to panic with a data abort trace.
In the end, same result.

`assert()` is only useful for debugging purposing to get an explicit trace on an specific unexpected event. Note that there are friendly backtrace tools in optee_os to analyse a data abort for example.

Comment (sander-visser) [reply core/crypto/aes-gcm-sw.c:46]: cpp check highlighted the issue of pointer arithmetic on potential null pointer. Id be open to other methods of hardening internal_aes_gcm_ghash_update...

File: core/arch/arm/mm/tee_pager.c (hunk 3 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c:541]: the if !fobj check existed since before - i just moved it up so that it actually can have an effect...

Comment (etienne-lms) [reply core/arch/arm/mm/tee_pager.c:541]: you're right, sorry. I missed the removed lines below.
Actually, should we panic() here, assert fobj is not NULL or should we simply remove the `if (!fobj) panic();` and let `size_t s = fobj->num_pages * SMALL_PAGE_SIZE;` trig a segfault caught by Core abort handler?

I would be in favor of the later and only remove lines 557..559.
@jenswi-linaro? @jforissier? and all?

Do we really want to assert all arguments in all internal or exported function?
There is no unpredictable behavior here on fault NULL reference: Core cleanly treats segfaults, possibly panicking when needed.

Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c:541]: static code analysis will highlight the risk or nullptr deref unless there is an assert that shows we are ok with not expecting null arg...

Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:541]: > Do we really want to assert all arguments in all internal or exported functions?

No, we don't want to add mindless asserts everywhere.

> static code analysis will highlight the risk or nullptr deref unless there is an assert that shows we are ok with not expecting null arg...

￼That is very unspecific. Which tool are we talking about? To me this sounds like asserts are used to silence warnings in that particular tool. Does this tool take only possible code paths into account like Coverity if configured correctly?

Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c:541]: The tool is Cppcheck 1.89 and it does not check for currently possible code paths, its focus is to make exported functions not crash.

Since the NULL check existed before my commit (i just moved it) i think it is better this way, but im fine with removing the null check and getting the panic as a side-effect if that is how other null ptr deref is managed.

File: core/arch/arm/mm/tee_mmu.c (hunk 2 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_mmu.c:374]: res != TEE_SUCCESS exists in 665 places since before

if (res) existed in 430 places since before

Iv changed to if (res)

Comment (sander-visser) [reply core/arch/arm/mm/tee_mmu.c:374]: there seems to be a mix in the codebase with about 60% res != TEE_SUCCESS and 40% if (res) - if changed anyways

File: core/crypto/aes-gcm-sw.c (hunk 4 above)
Comment (sander-visser) [reply core/crypto/aes-gcm-sw.c]: changed to optee coding style

File: core/arch/arm/kernel/ree_fs_ta.c (hunk 1 above)
Comment (sander-visser) [reply core/arch/arm/kernel/ree_fs_ta.c]: fixed

File: core/arch/arm/mm/tee_pager.c [hunk 6]
```diff
@@ -532,9 +532,14 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
 	uint32_t flags = 0;
 	size_t fobj_pgoffs = 0;
 	vaddr_t b = base;
-	size_t s = fobj->num_pages * SMALL_PAGE_SIZE;
+	size_t s;
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c:535]: Please assign a default value to `s`.

**Comment by jenswi-linaro** [comment]
For commits:
"core: fix TA REE backed secure storage memory leak"
"core: mm: fix MMU memory leak"
"core: rsa: Avoid NULL dereferencing in RSA trace messages"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jenswi-linaro** [comment]
Thanks for running the static analysis. Was all these issues originally found by the the cppcheck tool?
However, I'm not keen on adding cppcheck artefacts in the code like the one in the "core: added NULL arithmetic assert" commit.

File: core/arch/arm/mm/tee_pager.c [hunk 7]
```diff
@@ -532,9 +532,14 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
...
 	size_t s2 = 0;
 
+	if (!fobj)
+		panic();
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:539]: I would prefer an assertion here despite the fact there is an explicit (but useless) `panic()` in current implementation.

Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:539]: @etienne-lms, no strong opinion here, but I agree that you have a point.

**Comment by jenswi-linaro** [comment]
We've put some time into reviewing this, can we see this through please?

**Comment by jforissier** [comment]
@sander-visser ping?

File: core/arch/arm/mm/tee_pager.c (hunk 6 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c:535]: The default value is never used

Comment (etienne-lms) [reply core/arch/arm/mm/tee_pager.c:535]: you are right, but it is optee convention that expects all variables to get a default where defined. Compiler will take care of discarding useless default values.
See https://optee.readthedocs.io/en/latest/general/coding_standards.html, item 3.

**Comment by etienne-lms** [comment]
Thanks @sander-visser for the update.
Travis complains on some commit log issues that should be easy to address. See "continuous-integration/travis-ci/pr" above.

File: core/arch/arm/mm/tee_pager.c (hunk 6 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c]: fixed

File: core/arch/arm/mm/tee_pager.c (hunk 7 above)
Comment (sander-visser) [reply core/arch/arm/mm/tee_pager.c]: added a string to the panic (it was a paninc withotu a string too late in teh function before)

ok as is?

**Comment by sander-visser** [comment]
> 
> 
> Thanks @sander-visser for the update.
> Travis complains on some commit log issues that should be easy to address. See "continuous-integration/travis-ci/pr" above.

remaining travis warning is due to syntax mismatch in my commit name and signoff string.

Could you please review the last 2 commits (that have no review approval yet)

File: core/crypto/aes-gcm-sw.c [hunk 8]
```diff
@@ -38,14 +38,16 @@ void internal_aes_gcm_ghash_update(struct internal_aes_gcm_state *state,
 				   const void *head, const void *data,
 				   size_t num_blocks)
 {
-	size_t n = 0;
```
Comment (jenswi-linaro) [inline core/crypto/aes-gcm-sw.c:41]: Unrelated change

File: core/crypto/aes-gcm-sw.c [hunk 9]
```diff
@@ -38,14 +38,16 @@ void internal_aes_gcm_ghash_update(struct internal_aes_gcm_state *state,
...
 	if (head)
 		ghash_update_block(state, head);
 
-	for (n = 0; n < num_blocks; n++)
-		ghash_update_block(state,
-				   (uint8_t *)data + n * TEE_AES_BLOCK_SIZE);
+	if (data) {
+		size_t n = 0;
+
+		for (; n < num_blocks; n++)
```
Comment (jenswi-linaro) [inline core/crypto/aes-gcm-sw.c:47]: Unrelated change

**Comment by jenswi-linaro** [comment]
Please fix the author in all the commits to match the signed-of-by tag.

File: core/crypto/aes-gcm-sw.c [hunk 10]
```diff
@@ -38,14 +38,16 @@ void internal_aes_gcm_ghash_update(struct internal_aes_gcm_state *state,
...
+			ghash_update_block(state,
+					   (const uint8_t *)data + n * TEE_AES_BLOCK_SIZE);
```
Comment (etienne-lms) [inline core/crypto/aes-gcm-sw.c:49]: Must split this line that exceeds 80 charaters.

minor: I thinnk `const` is not mandated here.

You could simplify the patch using:
```patch
 	if (head)
 		ghash_update_block(state, head);
 
+	if (!data)
+		return;
 
 	for (n = 0; n < num_blocks; n++)
 		ghash_update_block(state,
 				   (uint8_t *)data + n * TEE_AES_BLOCK_SIZE);
```

File: core/arch/arm/mm/tee_pager.c [hunk 11]
```diff
@@ -532,9 +532,14 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
 	uint32_t flags = 0;
 	size_t fobj_pgoffs = 0;
 	vaddr_t b = base;
-	size_t s = fobj->num_pages * SMALL_PAGE_SIZE;
+	size_t s = 0;
 	size_t s2 = 0;
 
+	if (!fobj)
+		panic("null_fobj_arg");
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:539]: Yeah, an assert might even be preferable.

File: core/crypto/aes-gcm-sw.c (hunk 10 above)
Comment (sander-visser) [reply core/crypto/aes-gcm-sw.c:49]: I kept the const since there is no need to cast constness away.

I think it is good to avoid returns in the middle fo a function.

Id be willing to extract the data part to a static internal_aes_gcm_ghash_update_data_blocks() function in order to keep linelength down if it is not ok as is...

Comment (jenswi-linaro) [reply core/crypto/aes-gcm-sw.c:49]: > I think it is good to avoid returns in the middle of a function.

I think it's good to avoid complicated code in order to avoid returning in the middle of a function.
In this case it's not a big deal, but what @etienne-lms is suggesting is preferable and consistent with the style used in this project.

Comment (etienne-lms) [reply core/crypto/aes-gcm-sw.c:49]: ok fine.
For the 80 char per line convention, prefer:
```c
	if (data)
		for (n = 0; n < num_blocks; n++)
			ghash_update_block(state,
					   (const uint8_t *)data +
					   n * TEE_AES_BLOCK_SIZE);
```
Note: braces not required for the `if ()` block.

File: core/arch/arm/mm/tee_pager.c [hunk 12]
```diff
@@ -532,9 +532,13 @@ void tee_pager_add_core_area(vaddr_t base, enum tee_pager_area_type type,
 	uint32_t flags = 0;
 	size_t fobj_pgoffs = 0;
 	vaddr_t b = base;
-	size_t s = fobj->num_pages * SMALL_PAGE_SIZE;
+	size_t s = 0;
 	size_t s2 = 0;
 
+	assert(fobj);
+
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:539]: minor: remove this empty line.

File: core/crypto/aes-gcm-sw.c [hunk 13]
```diff
@@ -43,11 +43,11 @@ void internal_aes_gcm_ghash_update(struct internal_aes_gcm_state *state,
...
 		ghash_update_block(state, head);
 
-	if (data) {
+	if (data)
 		for (n = 0; n < num_blocks; n++)
 			ghash_update_block(state,
-					   (const uint8_t *)data + n * TEE_AES_BLOCK_SIZE);
-	}
+					   (const uint8_t *)data +
+					   n * TEE_AES_BLOCK_SIZE);
```
Comment (etienne-lms) [inline core/crypto/aes-gcm-sw.c:50]: This change belongs to commit "core: refactoring to avoid possible NULL_PTR arithmetics".

**Comment by jenswi-linaro** [comment]
Looks good now.

**Comment by jforissier** [comment]
@sander-visser thanks!
